File: jenkins-plugin/src/main/java/org/eazyportal/plugin/release/jenkins/ProjectDescriptorFactory.java
package org.eazyportal.plugin.release.jenkins;

import hudson.Extension;
import hudson.model.InvisibleAction;
import org.eazyportal.plugin.release.core.model.ProjectDescriptor;
import org.eazyportal.plugin.release.core.scm.ScmActions;
import org.eazyportal.plugin.release.gradle.project.GradleProjectActionsFactory;

import javax.inject.Inject;
import java.io.File;
import java.io.Serializable;

@Extension
public class ProjectDescriptorFactory extends InvisibleAction implements Serializable {

    @Inject
    private transient GradleProjectActionsFactory projectActionsFactory;

    private final transient org.eazyportal.plugin.release.core.ProjectDescriptorFactory delegate =
        new org.eazyportal.plugin.release.core.ProjectDescriptorFactory();

    public ProjectDescriptor create(File workingDir, ScmActions scmActions) {
        return delegate.create(projectActionsFactory, scmActions, workingDir);
    }

}

File: core/src/main/kotlin/org/eazyportal/plugin/release/core/ProjectDescriptorFactory.kt
package org.eazyportal.plugin.release.core

import org.eazyportal.plugin.release.core.model.Project
import org.eazyportal.plugin.release.core.model.ProjectDescriptor
import org.eazyportal.plugin.release.core.project.ProjectActionsFactory
import org.eazyportal.plugin.release.core.scm.ScmActions
import java.io.File

open class ProjectDescriptorFactory {

    fun create(
        projectActionsFactory: ProjectActionsFactory,
        scmActions: ScmActions,
        workingDir: File
    ): ProjectDescriptor {
        val rootProject: Project = workingDir.toProject(projectActionsFactory)

        val subProjects: List<Project> = scmActions.getSubmodules(workingDir)
            .map { workingDir.resolve(it) }
            .map { it.toProject(projectActionsFactory) }

        return ProjectDescriptor(
            allProjects = listOf(*subProjects.toTypedArray(), rootProject),
            rootProject = rootProject,
            subProjects = subProjects
        )
    }

    private fun File.toProject(projectActionsFactory: ProjectActionsFactory): Project =
        Project(this, projectActionsFactory.create(this))

}
